packages/mobx-state-tree/__tests__/core/object.test.ts

```diff
@@ -706,3 +707,115 @@ if (process.env.NODE_ENV === "development")
             "Invalid type definition for property 'x', it looks like you passed a function. Did you forget to invoke it, or did you intend to declare a view / action?"
         )
     })
+
+test("#993 - references should have a parent event when the parent has not been accesed before", () => {
+    const events: string[] = []
+
+    const L4 = types
+        .model("Todo", {
+            id: types.identifier,
+            finished: false
+        })
+        .actions(self => ({
+            toggle() {
+                self.finished = !self.finished
+            },
+            afterCreate() {
+                events.push("l4-ac")
+            },
+            afterAttach() {
+                events.push("l4-at")
+            }
+        }))
+
+    const L3 = types.model({ l4: L4 }).actions(self => ({
+        afterCreate() {
+            events.push("l3-ac")
+        },
+        afterAttach() {
+            events.push("l3-at")
+        }
+    }))
+
+    const L2 = types
+        .model({
+            l3: L3
+        })
+        .actions(self => ({
+            afterCreate() {
+                events.push("l2-ac")
+            },
+            afterAttach() {
+                events.push("l2-at")
+            }
+        }))
+
+    const L1 = types
+        .model({
+            l2: L2,
+            selected: types.reference(L4)
+        })
+        .actions(self => ({
+            afterCreate() {
+                events.push("l1-ac")
+            },
+            afterAttach() {
+                events.push("l1-at")
+            }
+        }))
+
+    const createL1 = () =>
+        L1.create({
+            l2: {
+                l3: {
+                    l4: {
+                        id: "11124091-11c1-4dda-b2ed-7dd6323491a5"
+                    }
+                }
+            },
+            selected: "11124091-11c1-4dda-b2ed-7dd6323491a5"
+        })
+
+    const expectedEvents = [
+        "l1-ac",
+        "l2-ac",
+        "l2-at",
+        "l3-ac",
+        "l3-at",
+        "l4-ac",
+        "l4-at",
+        "onSnapshot",
+        "onSnapshot"
+    ]
+
+    // test 1, real child first
+    {
+        const l1 = createL1()
+        onSnapshot(l1, () => {
+            events.push("onSnapshot")
+        })
+
+        l1.l2.l3.l4.toggle()
+        l1.selected.toggle()
+        expect(events).toEqual(expectedEvents)
+    }
+
+    // test 2, reference first
+    events.length = 0
+    {
+        const l1 = createL1()
+        onSnapshot(l1, () => {
+            events.push("onSnapshot")
+        })
+
+        l1.selected.toggle()
+        l1.l2.l3.l4.toggle()
+        expect(events).toEqual(expectedEvents)
+    }
+
+    // test 3, reference get parent should be available from the beginning
+    {
+        const l1 = createL1()
+        expect(getParent(l1.selected)).toBeTruthy()
```

Comment:
Note that this returns the parent of the selected's target, not (so L4's parent)

Reply:
yep, but if I change it to test against l3, then that reference will initialize it already, so the unit test wouldn't do much

Reply:
I'm not sure if I follow about the after attach and getParent thing...